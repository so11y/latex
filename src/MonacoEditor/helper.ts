import { Node } from "acorn"

export interface ValidateTypeResult {
  node: Node,
  through: boolean,
  message: string | null
}

export function throwNode(node: Node): ValidateTypeResult {
  return {
    node,
    through: true,
    message: null
  }
}

export function extractTokenAndNumbers(str: string) {
  // 匹配 "(1:22)" 格式的模式
  const regex = /\((\d+):(\d+)\)/;

  // 使用正则表达式进行匹配
  const match = str.match(regex);

  // 如果匹配成功，则返回提取的内容，包括两个数字
  if (match) {
    const extracted = {
      token: match[0], // 完整匹配的字符串
      firstNumber: parseInt(match[1]), // 第一个数字
      secondNumber: parseInt(match[2]) // 第二个数字
    };
    return extracted;
  } else {
    return null;
  }
}
