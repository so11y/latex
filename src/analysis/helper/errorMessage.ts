export function formatterError(_: TemplateStringsArray, ...arg: any[]) {
  const [message, ...temp] = arg;
  return temp.reduce((prev, next, index) => {
    return prev.replace(`$${index}`, next);
  }, message);
}

export const ErrorMessage = {
  Unknown: {
    UnKnownCall: "未知方法调用",
    UnknownSyntax: "未知语法",
    UnknownIdentifier: "未知字符",
    UnknownOperator: "不支持位运算等符号",
  },
  BinaryExpression: {
    EqNumberOrCall: "运算两边需要是数值或者函数调用表达式或者运算",
  },
  CallExpression: {
    ArgLengthExpect: "参数个数不匹配需要 $0 个参数",
  },
  ConditionalExpression: {
    OnlyLogical: `只能是使用三元和逻辑表达式符号 $0`,
    OnlyTrueCall: "真结果需要是函数调用并且不能再返回逻辑表达式了",
    OnlyFalseCall: "假结果需要是函数调用并且不能再返回逻辑表达式了",
    NotNestConditionalExpression:
      "不支持在直接嵌套Conditional函数和三元表达式,但是可以在搭配逻辑表达式使用",
  },
  Expression: {
    RootNotBool: "当前最终运算得出的可能是一个布尔值，请检查",
    NotCanRootAst: "最外层只能是函数调用或者运算或者三元表达在或者是数值",
  },
  Identifier: {
    CallKeyCode: "需要调用函数的方式写$0(word)",
  },
  Literal: {
    OnlyInCall: "字符串只能出现在调用里",
  },
  LogicalExpression: {
    OnlyLogical: "逻辑表达式两边需要是条件表达式 使用 && 或者 || 连接 ",
  },
  NumberLiteral: {
    NeedComposeBinary: "数字需要与运算符号配合",
  },
};
