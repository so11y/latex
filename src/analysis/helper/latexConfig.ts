import { ValidateGuardFalseMate } from "../types";
import { StringAccept } from "../callAccept/defaultAccept";

export interface ConfigRecord {
  [key: string]: LatexCallConfigType;
}

export interface LatexValidateCallAccept {
  key: string;
  describe: string;
  validate: (
    this: Omit<LatexCallConfigType["config"], "accept">,
    node: any,
    parent: any,
    index: number
  ) => true | ValidateGuardFalseMate;
}

export interface LatexCallConfigType {
  name: string;
  alias: string;
  astName?: string;
  config: {
    accept: Array<LatexValidateCallAccept>;
  };
}

export class LatexConfig {
  LatexCallConfig = {
    Count: cratedDefaultAccept("Count", "数据量"),
    Mean: cratedDefaultAccept("Mean", "平均值"),
    Median: cratedDefaultAccept("Median", "中位数"),
    Absolute: cratedDefaultAccept("Absolute", "绝对值"),
    Min: cratedDefaultAccept("Min", "最小值"),
    Max: cratedDefaultAccept("Max", "最大值"),
    First: cratedDefaultAccept("First", "第一个点"),
    Last: cratedDefaultAccept("Last", "最后一个点"),
  } satisfies ConfigRecord;

  get LatexNames() {
    return Object.values(this.AllLatex).map((latex) => latex.name);
  }

  get AllLatex() {
    return {
      ...this.LatexCallConfig,
      ...macroLatexCallConfig,
    };
  }
}

export const macroLatexCallConfig = {
  Conditional: {
    name: "Conditional",
    alias: "三元表达式",
  },
};

function cratedDefaultAccept(name: string, alias: string) {
  return {
    name,
    alias,
    config: {
      accept: [StringAccept],
    },
  };
}
