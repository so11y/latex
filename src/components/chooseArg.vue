<template>
  <n-space>
    <n-tag
      type="success"
      v-for="(tag, index) in modelValue"
      :key="tag.key"
      @close="handleClose(index)"
      closable
    >
      <n-dropdown
        trigger="hover"
        label-field="describe"
        :options="(AllLatexCallAccept as any[])"
        @select="handleSelect($event, index)"
      >
        {{ tag.describe }}
      </n-dropdown>
    </n-tag>
  </n-space>
</template>

<script lang="ts" setup>
import { NSpace, NTag, NDropdown } from "naive-ui";
import { LatexValidateAccept } from "../analysis/helper/latexConfig";
import { AllLatexCallAccept } from "../analysis/callAccept/defaultAccept";

const props = defineProps<{ modelValue: Array<LatexValidateAccept> }>();

function handleSelect(key: string, index: number) {
  const accept = AllLatexCallAccept.find((accept) => accept.key === key)!;
  props.modelValue[index] = accept;
}

function handleClose(index: number) {
  props.modelValue.splice(index, 1);
}
</script>
