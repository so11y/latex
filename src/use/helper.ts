import { LatexComponent } from "../components/types"
import { ComponentName } from "./constName";


export function inspectType(l: LatexComponent, name: ComponentName) {
  return l.componentName === name || l.latexName === name
}
