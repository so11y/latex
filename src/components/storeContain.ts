import {
  ComputedRef,
  defineComponent,
  InjectionKey,
  provide,
  unref,
  reactive,
  watch,
  ref,
  nextTick,
  Ref,
} from "vue";
import type { ContainProvide, ContainLatexComponentStore } from "./types";
import { decode } from "../convert";

export type StoreValue = ComputedRef<ContainLatexComponentStore>;
export type Store = Record<string, StoreValue>;
export interface StoreContainProvide {
  getIncreaseUid(peek?:    boolean): string;
  getComponent(id: string): ContainLatexComponentStore;
  getComponentOrginl(id: string): StoreValue;
  getActiveId(): string;
  addComponent(i: StoreValue): void;
  removeComponent(id: string): void;
  setActiveId(id: string): void;
  putActiveChildren(
    latexName: string,
    alias: string,
    _componentName?: string,
    data?: Record<string, any>,
    id?: string | null
  ): void;
  toStringLatex: Ref<string>;
  encode(id?: string): Record<string, any>;
  decode(v: Record<string, any>): void;
}
function generateCombinedId() {
  const timestamp = new Date().getTime(); // 获取当前时间戳
  const randomId = Math.random().toString(36).substring(2); // 生成随机字符串
  const combinedId = randomId + '_' + timestamp; // 结合随机字符串和时间戳
  return combinedId;
}
// export const StoreContainProvideKey =  "StoreContainProvideKey"
export const StoreContainProvideKey = Symbol(
  "StoreContainProvideKey"
) as InjectionKey<StoreContainProvide>;

function handleToString(store: Store) {
  let latexString = ref("");
  watch(
    store,
    async () => {
      await nextTick();
      const root = unref(store[-1]);
      latexString.value = root ? root.toStringLatex() : "";
    },
    {
      deep: true,
    }
  );
  return latexString;
}

export default defineComponent({
  setup(_, { slots }) {
    const store = reactive<Store>({});
    let increaseId = generateCombinedId();
    let activeId: string = "-2";

    const toStringLatex = handleToString(store);

    const actions = {
      store,
      getIncreaseUid(peek = false) {
        if (peek) return increaseId;
        increaseId = generateCombinedId();
        return increaseId;
      },
      putActiveChildren(
        latexName,
        alias: string,
        _componentName?: string,
        data = {},
        id: string | null = null
      ) {
        if (activeId !== null) {
          const component = actions.getComponent(activeId);
          if (component) {
            const shareCtx = component as ContainProvide;
            const impl = {
              id: id ?? actions.getIncreaseUid(),
              componentName: _componentName ? _componentName : "Contain",
              latexName,
              parentId: activeId,
              alias,
              data,
            };
            if (shareCtx.onCanAddChildren?.() === false) {
              return;
            }
            if (shareCtx.onWaylayAddChildren?.(impl) === true) {
              return;
            }
            if (shareCtx.componentName === "FakeBlock") {
              const fakeBlockParent = actions.getComponent(shareCtx.parentId)
              const index = fakeBlockParent.children.value.findIndex(child => child.id === shareCtx.id)
              if (index > -1) {
                fakeBlockParent.children.value[index] = impl;
              }
              Reflect.deleteProperty(store, shareCtx.id);
            } else {
              shareCtx.children.value.push(impl);
            }
          }
        }
      },
      getComponentOrginl(i) {
        return store[i]
      },
      getComponent(i) {
        return unref(store[i]);
      },
      addComponent(i) {
        const { value } = i;
        store[value.id] = i;
      },
      removeComponent(i) {
        const self = actions.getComponent(i);
        if (!self || self.parentId === undefined) return;
        const parent = actions.getComponent(self.parentId) as ContainProvide;
        Reflect.deleteProperty(store, i);
        if (parent) {
          const index = parent.children.value.findIndex(
            (child) => child.id === self.id
          );
          const impl = {
            id: actions.getIncreaseUid(),
            componentName: "FakeBlock",
            latexName: "FakeBlock",
            parentId: parent.id,
            alias: "",
            data: {},
          };
          parent.children.value[index] = impl;
        }
      },
      setActiveId(i) {
        activeId = i;
      },
      getActiveId() {
        return activeId;
      },
      toStringLatex,
      encode(id = "-1") {
        const s = actions.getComponent(id).encode();
        console.log(s);
        return s;
      },
      decode(v: Record<string, any>) {
        if (v.config.bracket) {
          let current_bracket = v.config;
          while (current_bracket.bracket) {
            current_bracket = current_bracket.bracket;
          }
          current_bracket.parentId = activeId;
        } else {
          v.config.parentId = activeId;
        }
        decode(v, actions);
      },
    } as StoreContainProvide;
    provide(StoreContainProvideKey, actions);
    return () => {
      return slots.default!(actions);
    };
  },
});
