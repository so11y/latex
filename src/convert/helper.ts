import { omit } from "lodash-es";
import { ContainLatexComponentStore, ignoredPropsKey } from "../components/types";
import { StoreContainProvide } from "../components/storeContain";

export function buildConfig(v: ContainLatexComponentStore) {
  const ignorePackOptionKeys = [
    ...ignoredPropsKey,
    "children",
    "hasChildrenBobbleColumn",
    "isLast",
    "isFiled",
    "dept",
  ]
  if (v.data && Object.keys(v.data).length === 0) {
    ignorePackOptionKeys.push("data");
  }
  const _props = omit(v, ignorePackOptionKeys);
  return _props;
}

export function encodeBracket(v: ContainLatexComponentStore, s: StoreContainProvide) {
  const [first] = v.children.value;
  const nest = s.getComponent(first.id);
  nest.bracket = buildConfig(v);
  return nest;
}
