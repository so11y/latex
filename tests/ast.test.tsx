import ConditionalExpressionAst from "./features/ConditionalExpression.json";
import BinaryExpression from "./features/BinaryExpression.json";
import NestBracket from "./features/NestBracket.json";
import { assert, describe, expect, test } from "vitest";
import { createSimpleApp } from "./helper/createApp";
import { StoreContainProvide } from "../src/components/storeContain";
import { nextTick } from "vue";

async function astTest(json: Record<string, any>) {
  let store: StoreContainProvide;
  const app = createSimpleApp((s) => {
    store = s;
    return null;
  });
  //@ts-ignore
  if (!store) return;
  store.setActiveId("-1");
  await store.decode(json);
  await nextTick();
  expect(store.encode()).toEqual(json);
  expect(app.html()).toMatchSnapshot();
}

describe("ast ", () => {
  test.each([
    {
      ast: ConditionalExpressionAst,
      name: "ConditionalExpressionAst",
    },
    {
      ast: BinaryExpression,
      name: "BinaryExpression",
    },
    {
      ast: NestBracket,
      name: "NestBracket",
    },
  ])("ast($name)", async (a, b) => {
    await astTest(a.ast);
  });
});
