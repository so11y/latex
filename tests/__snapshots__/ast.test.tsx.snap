// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ast  > ConditionalExpressionAst  1`] = `
"<div data-v-73788115="" data-id="-1" class="latex-contain column column_contain">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap column">
    <div data-v-73788115="" data-id="1" class="latex-contain column column_contain">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Conditional<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap column">
        <div data-v-73788115="" data-id="2" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Eq<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="3">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Absolute<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="4">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> aa
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="5">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Median<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="6">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> bb
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
        <div data-v-73788115="" data-id="7" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Div<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="8">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="9">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> 3
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="10">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Median<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="11">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> cc
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
        <div data-v-73788115="" data-id="12" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Plus<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="13">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Last<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="14">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> dd
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="15">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="16">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> 1
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
      </div><span><!--v-if--></span><span><!--v-if--></span><span><!--v-if--></span>
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;

exports[`ast  > NestBracket 1`] = `
"<div data-v-73788115="" class="latex-contain" data-id="-1">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap">
    <div data-v-73788115="" class="latex-contain" data-id="1">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap">
        <div data-v-73788115="" class="latex-contain" data-id="2">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap">
            <div data-v-73788115="" class="latex-contain" data-id="3">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="4">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
                  </div>
                  <div data-v-73788115="" class="latex_gap">
                    <div data-v-73788115="" class="latex-contain" data-id="5">
                      <div data-v-73788115="" class="latex-contain__title">
                        <div data-v-73788115="" class="remove">x</div> 3
                        <!--v-if-->
                      </div>
                      <!--v-if-->
                    </div>
                  </div><span><!--v-if--></span>
                  <div data-v-73788115="" class="bracket bracket-end"> ) </div>
                </div>
              </div>
              <!--v-if-->
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div>
          <!--v-if-->
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
      </div>
      <!--v-if-->
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;

exports[`ast  > ast('BinaryExpression') 1`] = `
"<div data-v-73788115="" data-id="-1" class="latex-contain column column_contain">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap column">
    <div data-v-73788115="" data-id="1" class="latex-contain column column_contain">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Plus<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap column">
        <div data-v-73788115="" class="latex-contain" data-id="2">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap">
            <div data-v-73788115="" class="latex-contain" data-id="3">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> 1
                <!--v-if-->
              </div>
              <!--v-if-->
            </div>
          </div><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
        <div data-v-73788115="" class="latex-contain" data-id="4">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap">
            <div data-v-73788115="" class="latex-contain" data-id="5">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> 2
                <!--v-if-->
              </div>
              <!--v-if-->
            </div>
          </div><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
      </div><span><!--v-if--></span><span><!--v-if--></span>
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;

exports[`ast  > ast('ConditionalExpressionAst') 1`] = `
"<div data-v-73788115="" data-id="-1" class="latex-contain column column_contain">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap column">
    <div data-v-73788115="" data-id="1" class="latex-contain column column_contain">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Conditional<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap column">
        <div data-v-73788115="" data-id="2" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Eq<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="3">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Absolute<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="4">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> aa
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="5">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Median<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="6">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> bb
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
        <div data-v-73788115="" data-id="7" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Div<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="8">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="9">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> 3
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="10">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Median<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="11">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> cc
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
        <div data-v-73788115="" data-id="12" class="latex-contain column column_contain">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Plus<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap column">
            <div data-v-73788115="" class="latex-contain" data-id="13">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Last<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="14">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> dd
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
            <div data-v-73788115="" class="latex-contain" data-id="15">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="16">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> 1
                    <!--v-if-->
                  </div>
                  <!--v-if-->
                </div>
              </div><span><!--v-if--></span>
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div><span><!--v-if--></span><span><!--v-if--></span>
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
      </div><span><!--v-if--></span><span><!--v-if--></span><span><!--v-if--></span>
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;

exports[`ast  > ast('NestBracket') 1`] = `
"<div data-v-73788115="" class="latex-contain" data-id="-1">
  <div data-v-73788115="" style="color: #CC99FF;" class="latex-contain__title">
    <div data-v-73788115="" class="remove">x</div> Root<span data-v-73788115="" class="bracket">(</span>
  </div>
  <div data-v-73788115="" class="latex_gap">
    <div data-v-73788115="" class="latex-contain" data-id="1">
      <div data-v-73788115="" style="color: #66CCFF;" class="latex-contain__title">
        <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
      </div>
      <div data-v-73788115="" class="latex_gap">
        <div data-v-73788115="" class="latex-contain" data-id="2">
          <div data-v-73788115="" style="color: #003366;" class="latex-contain__title">
            <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
          </div>
          <div data-v-73788115="" class="latex_gap">
            <div data-v-73788115="" class="latex-contain" data-id="3">
              <div data-v-73788115="" style="color: #330066;" class="latex-contain__title">
                <div data-v-73788115="" class="remove">x</div> Bracket<span data-v-73788115="" class="bracket">(</span>
              </div>
              <div data-v-73788115="" class="latex_gap">
                <div data-v-73788115="" class="latex-contain" data-id="4">
                  <div data-v-73788115="" class="latex-contain__title">
                    <div data-v-73788115="" class="remove">x</div> Number<span data-v-73788115="" class="bracket">(</span>
                  </div>
                  <div data-v-73788115="" class="latex_gap">
                    <div data-v-73788115="" class="latex-contain" data-id="5">
                      <div data-v-73788115="" class="latex-contain__title">
                        <div data-v-73788115="" class="remove">x</div> 3
                        <!--v-if-->
                      </div>
                      <!--v-if-->
                    </div>
                  </div><span><!--v-if--></span>
                  <div data-v-73788115="" class="bracket bracket-end"> ) </div>
                </div>
              </div>
              <!--v-if-->
              <div data-v-73788115="" style="color: #330066;" class="bracket bracket-end"> ) </div>
            </div>
          </div>
          <!--v-if-->
          <div data-v-73788115="" style="color: #003366;" class="bracket bracket-end"> ) </div>
        </div>
      </div>
      <!--v-if-->
      <div data-v-73788115="" style="color: #66CCFF;" class="bracket bracket-end"> ) </div>
    </div>
  </div>
  <!--v-if-->
  <div data-v-73788115="" style="color: #CC99FF;" class="bracket bracket-end"> ) </div>
</div>"
`;
